### PR TITLE
[2186] Scaling utils

### LINKED ISSUE #2186: Create scaling plots for compute time proposal
### Describe the task. It can be a feature, documentation, etc.

For Jupiter and Santis proposals, we need to show how the code scales to many nodes.

- Up to 256 nodes (run 1,2,4,8,16,32,64,128,256)
- Should capture at least 100 training steps (10 minutes is expected to be sufficient)
- Configurations: jepa and forecasting
- Track throughput (samples/sec) for proposal, analyze loss vs #nodes (not for proposal).

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://gitlab.jsc.fz-juelich.de/hedgedoc/e0bUp55nSSWpXTYzSuj5QQ

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [x] performance

### PR DESCRIPTION
**Pull request by florianscheidl**
## Description

Contains utility we used for scaling experiments, namely to capture elapsed time metrics in the trainer and run_train.
Small code modifications became necessary as we trained on more than 16 nodes. Specifically:
- Lower bound on warmup, cooldown, and decay steps in lr_scheduler
- Lower bound on beta2 used in Adam optimizer (negative values are illegal). Set to 0.9 now, based on Internet suggestion, but open to suggestions.
- Lower bound on len_per_rank when setting mini_epoch_base (avoiding zero-division error for small configs).

The data extraction and plot generation script are addressed in https://gitlab.jsc.fz-juelich.de/esde/WeatherGenerator-private/-/merge_requests/180.

Example plots and tables: 

jupiter_weak_scaling_table
jupiter_weak_scaling

## Issue Number

Fixes #2186

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -558,6 +568,19 @@ def train(self, mini_epoch):
 
         self.dataset.advance()
 
+        if is_root():
+            total_training_time = time.time() - self.t_training_start
+            self.train_logger.log_metrics(
+                "train",
+                {
+                    "completed_mini_epoch": mini_epoch,
+                    "elapsed_time_mini_epoch": total_training_time,
```
Comment (ekouts) [inline src/weathergen/train/trainer.py]: It's a minor thing, but for me the name of the metric (`elapsed_time_mini_epoch`) is misleading. I would assume it measures the time for the mini epoch, but it measures the overall training time until this mini epoch ended. Could we find a better name? `cumulative_time_mini_epoch_seconds` is too long? Mostly because in some other metrics we have the unit in the end.

**Review by ekouts — CHANGES_REQUESTED** [review]

File: src/weathergen/train/trainer.py (hunk 1 above)
Comment (florianscheidl) [reply src/weathergen/train/trainer.py]: Good point. What do you think of `training_time_after_mini_epoch_seconds `. It's long, but maybe clearer?

Comment (ekouts) [reply src/weathergen/train/trainer.py]: Yes, also fine by me.

**Review by ekouts — APPROVED** [review]

**Comment by florianscheidl** [comment]
@clessig, anything blocking?

File: packages/performance/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# WeatherGenerator Performance Analysis Tools
```
Comment (clessig) [inline packages/performance/README.md]: Can this directory be moved to the private repo

File: src/weathergen/utils/train_logger.py [hunk 3]
```diff
@@ -101,9 +101,10 @@ def add_logs(
         stddev_all: dict,
         avg_loss: list[float] = None,
         lr: float = None,
+        elapsed_training_time_seconds: float | None = None,
```
Comment (clessig) [inline src/weathergen/utils/train_logger.py:104]: Can you open a PR that we change plot_train so that we can plot losses against training time.

File: src/weathergen/run_train.py [hunk 4]
```diff
@@ -188,7 +189,7 @@ def run_train(args):
     trainer = Trainer(cf.train_logging)
 
     try:
-        trainer.run(cf, devices)
+        trainer.run(cf, devices, t_start=t_start)
```
Comment (clessig) [inline src/weathergen/run_train.py]: Why can t_start not been done in `trainer.py`?

**Review by clessig — COMMENTED** [review]
See detailed comments.

File: src/weathergen/run_train.py (hunk 4 above)
Comment (florianscheidl) [reply src/weathergen/run_train.py]: We chose so to capture the time from the beginning of run_train onwards.

Comment (clessig) [reply src/weathergen/run_train.py]: From a SWE point of view it would be much better to keep it contained in Trainer.

**Comment by florianscheidl** [comment]
I've implemented the changes discussed here and in https://github.com/ecmwf/WeatherGenerator/pull/2501. The latter will only have the plotting changes.

**Review by clessig — APPROVED** [review]
